source/common/network/dns_resolver/dns_factory_util.cc

```diff
@@ -79,8 +79,11 @@ void handleLegacyDnsResolverData(
 Network::DnsResolverFactory& createDnsResolverFactoryFromTypedConfig(
     const envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config) {
   ENVOY_LOG_MISC(debug, "create DNS resolver type: {}", typed_dns_resolver_config.name());
-  return Config::Utility::getAndCheckFactory<Network::DnsResolverFactory>(
-      typed_dns_resolver_config);
+  auto& factory =
+      Config::Utility::getAndCheckFactory<Network::DnsResolverFactory>(typed_dns_resolver_config);
+  // Perform factory initialization.
+  factory.init();
```

Comment:
If you move this code into the process_wide.cc you will not need complexity with the `ares_library_initialized_` in the factory. You already put the code for clean-up there.

Reply:
The reason we do factory.init() here is that only if the factory is c-ares, it will call ares_library_init() and set ares_library_initialized_ to be true.  If c-ares Resolver is never created, then  ares_library_init() will never be called, and ares_library_initialized_  stay false.   This provide a path to never call these ares_library_init()/ares_library_cleanup() functions if proper DNS resolvers are configured.

If we do this in process_wide(), there is no context there to check whether c-ares resolver is configured. That will ends up ares_library_init() get called even c-ares DNS resolver is never created.